### ISSUE TITLE
[OJS] The current role does not have access to this operation -message needs more details on how to acquire the needed role

### ISSUE THREAD

**Issue by ajnyga** [pkp/pkp-lib#2297]
A typical problem in multi-journal installations is that a user has registered to journal A and later want's to submit something to journal B as well. However, usually this is not possible, because the user does not have the author role in journal B.

I made a feature request about how OJS should handle this situation. http://forum.pkp.sfu.ca/t/ojs3-missing-author-roles-in-multi-journal-installations/28710

In my opinion having a simple button to enable the author role in that message would be ideal. However, I developed a quick fix where the message text has an additional link leading to the user profile/roles page. This will probably help in most cases. PR in a minute.

**Comment by ajnyga** [comment]
This is a translation related solution, so tagging @mtub here.

**Comment by asmecher** [comment]
I'll continue the forum conversation about the "ideal" solution in a sec, but one expedient option might be to have this specific case use a different (more useful) locale key. That way you wouldn't need to modify the general-purpose one in a way that would be misleading for roles that can't be self-registered.

**Comment by ajnyga** [comment]
Thanks, I will check `setAdvice`. I am using the above pr in our local installation and it seems to have stopped the quite frequent questions concerning the issue above. However, the hard coded url would have to change in any case like you noticed.

**Comment by ajnyga** [comment]
sorry @asmecher but the authorization system in OJS is just too complicated for me to understand. Also I could not quite understand how the setAdvice is supposed to work :S

**Comment by ajnyga** [comment]
@asmecher @NateWr 
In multijournals installations this is something that is creating a lot of problems. Users registered in another journal can log in to a journal but they do not have permissions to for example send submissions. We get a lot of these and users seem to be lost as there is no clear instructions how to add the required role. Just an error message saying that "you can not do that".

I think that this could be fixed by making it easier for users to acquire the author role when it is missing. Can you check my last comment from September at some point so that I can move forward in implementing the new feature mentioned above.

**Comment by NateWr** [comment]
Personally, I don't think users should need to "acquire" the Author role at all. Unless a journal has shut down non-commissioned submissions (ie - by closing all registrations), any user should be able to initiate a submission.

If they already have an Author role in any context, and they initiate a submission in another context, we should just take them to the submission page and make them select an Author-ish role as the first item in the submission form.

**Comment by ajnyga** [comment]
Would this cause problems with things like GDPR? When I suggested that journal manager should be able to add roles for users from other journals, @bozana raised the question of GDPR and the fact that the user has to register the role herself for the related legistlation to be fulfilled.

Or do you think that doing a submission can be regarded as a contract as such? I mean automatically add the role when the user starts to send a manuscript. Usability wise it would be a perfect solution imho.

**Comment by NateWr** [comment]
I am not a lawyer, but if the journals are all part of the same site (eg - `journal.fi/...`) I think they would all fall under the same terms. In the same way that an admin can assign a user a new role, so could we do this automatically, I would think.

Even if the journals are separate, though, it's hard to see what the difference would be. When a user makes a submission, to comply with GDPR the site would need to specify how their data is stored/used. If a role ID is one part of their submission data, it doesn't seem unique in that sense.

**Comment by ajnyga** [comment]
Just tagging the other issue here: https://github.com/pkp/pkp-lib/issues/3022
And the discussion: https://forum.pkp.sfu.ca/t/ojs3-editing-user-roles-in-multi-journal-installations/28711

As you can see there, site admin can assign roles freely, but the journal managers can not. But this could just be an issue when someone else is doing something in your behalf. 

Actually Alec comments to your suggestion here: https://forum.pkp.sfu.ca/t/ojs3-missing-author-roles-in-multi-journal-installations/28710/2
"We had a solution for this way back in OCS – the user would simply gain the required role if possible when the submission link was followed – but that approach won’t work with OJS 3.x/OMP because there may now be several submitting roles and we’ll need the author to choose one."

This lead me to think of a solution where clicking the "Make a new submission" which now just gives you a text saying " The current role does not have access to this operation. " would instead give you an easy way of adding an author role. And that is what was pursued above.

**Comment by NateWr** [comment]
> This lead me to think of a solution where clicking the "Make a new submission" which now just gives you a text saying " The current role does not have access to this operation. " would instead give you an easy way of adding an author role. And that is what was pursued above.

I'd rather just get them started on their submission. They can select an Author role as part of the first form they submit in the process. Something like "Submitting as" with a select of Author roles (or, if only one `ROLE_ID_AUTHOR` role exists, we can include a hidden field).

**Comment by ajnyga** [comment]
That would probably mean changing some of the access policies in the form, but I do think that is a good idea to add those to the form.

If @asmecher agrees I can try doing this during the weekend (while eating mämmi).

**Comment by bozana** [comment]
Hi all, just a comment regarding data privacy: I think the problem is when someone else (e.g. journal manager or editor) does something for another person/user (and the user does not know about that i.e. does not agree i.e. does not have a chance to agree/disagree). I.e. it would need explicitly agreement of the user. If a user him/herself would like to do something, that is OK. 
Thus, for example, the problem would/could for example be:
-- when a journal manager or an editor creates a new user (e.g. reviewer) or enrolls a user for a journal without his/her agreement
-- when a journal manager or editor from one journal can see the (user) data that is actually registered for another journal. 
It could be that all journals in that installation are separate and do not have anything to do with each other. It could be also the other way around, that all journals belong to one instance. In both use cases there could be both: only one main domain for all journals or each journal could have its own domain. E.g. a university library could host all university journals under one domain, however those journals could be (also only partly) separate entities and would not have to do anything with each other.
We consider and would like to support both/all use cases, that's why we separate the data per default per journal, but still allow the other use case. We should probably improve all that. However, in both cases, the users should exactly know what is happening (with their data) and how, before they register and start to use the system i.e. agree with that way/kind of their data use.
It is normal that one special user (admin) can see everything, I believe. It is similar to a system admin, that can see all the data in the DB. This is a special kind of role and it should be at minimum i.e. not many users should have that kind of access.
-- Just my/me thinking...

**Comment by asmecher** [comment]
> I am not a lawyer, but if the journals are all part of the same site (eg - journal.fi/...) I think they would all fall under the same terms.

Yes, but probably not for service providers using OJS as SaaS, where journals might appear at different domains with no indication of shared platform.

> I'd rather just get them started on their submission. They can select an Author role as part of the first form they submit in the process. Something like "Submitting as" with a select of Author roles (or, if only one ROLE_ID_AUTHOR role exists, we can include a hidden field).

Yes, but for journals using the default role configuration (99% I'd guess) there will really only be a single author role allowing self-registration, so it's a moot point. I think we can choose the author role when there's only one, and prompt for the few remaining cases, and if the latter group finds it unworkable we can ask them for suggestions for improvement.

**Comment by asmecher** [comment]
> ...while eating mämmi...

I had to look that up -- happy Easter!

**Comment by ajnyga** [comment]
So to sum up:
User without the author role selects "Make a new submission"
=> we check if the context has only one author role, if it does we automatically add the role
=> if there are more than one role, then add the role selection to the first stage of the submission role: "You need to select one of the author roles available for this context".

**Comment by asmecher** [comment]
Yes, sounds great. This might mean excluding the first step of the submission wizard from requiring the author role, but OTOH we've done something like that elsewhere so there might be a pattern to follow.

**Comment by ajnyga** [comment]
But @asmecher do check the AuthorizationDecisionManager code I mention above.

Because if you look at line https://github.com/pkp/pkp-lib/blob/master/classes/security/authorization/AuthorizationDecisionManager.inc.php#L129
And line https://github.com/pkp/pkp-lib/blob/master/classes/security/authorization/AuthorizationDecisionManager.inc.php#L124

I can not understand how the `if` on line 129 could **ever** be true?

**Comment by ajnyga** [comment]
Hi @asmecher and @NateWr 

I know this is what we discussed.

But while doing this change, I found this function that is already used in several places `$userGroupDao->getDefaultByRoleId`. Like you probably guessed, it fetches a single default author role from a given context.

This seemed so perfect, that I created a pr based on that: https://github.com/pkp/pkp-lib/pull/3542

If you still feel like we should give the authors a choice if there are several author roles, then I can do the additional changes needed. It is actually not much.

Also I funny thing I noticed after I changed the policy so that users with no roles could access the submission form: there was already a secondary check there. When the author role was missing, the form showed this error: https://github.com/pkp/pkp-lib/blob/master/locale/en_US/submission.xml#L16

If you think that additional changes are not needed, then I do not think that locale key is needed anymore. But if you think we need the role selection, then I would use that same key just by removing the last sentence. What do you think.

Also Alec, check my last comment as well. I am fairly sure that is a bug although this pr is not affected anymore.

**Comment by NateWr** [comment]
That sounds good to me. I won't be able to code review this until at least next week but happy to take a look then.

**Comment by ajnyga** [comment]
No hurry from my part!

**Comment by asmecher** [comment]
Thanks, @ajnyga and @NateWr -- and apologies as always for being slow to respond. Nate, let me know if you're not sure about something.

**Comment by ajnyga** [comment]
Thanks! @asmecher, did you check the bug here: https://github.com/pkp/pkp-lib/issues/2297#issuecomment-377126319 (not related to the pr anymore, but will most likely affect something else later)

**Comment by asmecher** [comment]
> I can not understand how the if on line 129 could ever be true?

Check out https://github.com/pkp/pkp-lib/blob/master/classes/security/authorization/AuthorizationDecisionManager.inc.php#L228 -- the `$callOnDeny` is passed into this by reference, thus gets set in the calling scope too.

**Comment by ajnyga** [comment]
Thanks!
I wonder why the example here did not work for me: https://github.com/pkp/pkp-lib/issues/2297#issuecomment-329101825 ? (not really important if you know that the code is correct at the moment)

**Comment by ajnyga** [comment]
I give up! Hopefully I do not have to work with the policies any time soon 🤣

**Comment by ajnyga** [comment]
@NateWr nevermind the conversation above, the pr above is ready for review whenever you have the time.

**Comment by ajnyga** [comment]
Ok, so in multijournal installations. Even if there is a site wide privacy policy, I guess it could be possible that a single journal has some different journal policies, right? Also joining the workflow of a new journal as an author will mean that your personal data is handled by different people than before, right?

@jmacgreg what do you think? 

I think it would be easy to add a solution where entering the submission form without an author role will show you a pull down menu with all the author roles and a text containing a link to the journal privacy policy. Maybe even have a checkbox for checking that you have read the statement (the same consent you would have in the registration form). This would probably be the safest way of filling the GDPR requirements and not even that difficult to handle.

**Comment by NateWr** [comment]
> I think it would be easy to add a solution where entering the submission form without an author role will show you a pull down menu with all the author roles and a text containing a link to the journal privacy policy

Use radio buttons so the user can see all the options at the start, and use `$userGroupDao->getDefaultByRoleId` to pre-select an option.

> Is there already some handler that shows the journal privacy policy? Maybe something that opens the policy to a modal window?

Not that I know of.

If this is necessary for GDPR then so be it. But I'd just like to encourage us to be a bit conservative about what extra conditions, checkboxes, and agreements we require a user to get through in the submission process. We should only ever ask someone to agree to terms when we absolutely have to, and we should go out of our way to prevent scenarios where someone is agreeing to the same terms twice.

**Comment by ajnyga** [comment]
Ok, I will do a pr that adds the selection you mentioned. The privacy statement text I will leave out for now. Let's see what happens with the registration form first regarding GDRP.

**Comment by jmacgreg** [comment]
As a general practice I'd suggest always at least including a link to the journal's privacy policy whenever a user is updating their data (and that should maybe be filed as a separate issue). I don't think there's a need for an explicit consent action as part of enabling a role, if someone has already registered. 

There should also be a means to consent to data processing as part of the submission process, including consent  from other authors (ie. non-users in the system), but that can be handled with a combination of submission checklist + consent form templates (to be uploaded as a submission file). 

To sum up: I think explicit consent at time of registration; and explicit consent at time of submission; and, eventually, the ready availability of the journal's privacy policy anywhere where users update their accounts, should be fine.

**Comment by ajnyga** [comment]
Ok so the privacy statement is actually already on the submission page...

**Comment by asmecher** [comment]
@ajnyga, if you're stuck on a particular aspect of the policy framework and want to walk me through a scenario (e.g. give me a database dump and set of steps to reproduce an unexpected outcome), I'd be happy to take a look. It's a pain to debug.

**Comment by ajnyga** [comment]
@asmecher the policy aspect is really not a problem here anymore, it was just a side track because I thought that there was a coding error there.

The discussion here has been really just about whether the enrollment should be automatic or not. We actually came to a conclusion that it should not be automatic and will be doing basically this: https://github.com/pkp/pkp-lib/pull/3542#discussion_r181406374

**Comment by asmecher** [comment]
Sure -- but if you find yourself bashing your head on this sometime later, the offer is available.

**Comment by ajnyga** [comment]
Hi, @NateWr @asmecher 

This is the new version of the pr: https://github.com/pkp/pkp-lib/pull/3542


At the moment this is happening in master branch:
- if editor submits a manuscript and only has manager roles, the editor will be added to the contributors list of that submission, but not as an author in the stage assignments. This is probably what most editors expect when they send things like editorials.
- If we would add the author role selection also for the users with only manager roles, then that role would be carried on to the stage assignments. But, I think that some editors still want to submit papers as editors

The pr will support the following scenarios:
- Case 1: users without any roles: show selection of available author roles to choose from. Tell the user she has to enroll.
- Case 2: user with existing author role: if one existing author role, add hidden variable for the userGroupId, if several existing author roles add a selection to choose from.
- Case 3: manager users _with_ existing author roles: show selection of both existing manager roles and available author roles and let the editor choose.
- Case 4: manager users _without_ existing author roles: show selection of both existing manager roles and available author roles and let editor choose. If she chooses an author role, enroll the manager also as an author.

This would solve the situations where editors want to submit to their own journal and receive a peer-review for their paper. In these situations they of course choose an author role. In other cases, like when sending in editorials etc. they can choose a managerial role. 

Question: do editors need some guidance for choosing between the manager and author roles?

This is maybe something you want to discuss with others as well? I have been thinking about different user scenarios and I think that this is all very logical and we do not have the "no role" dead end anymore.

**Comment by ajnyga** [comment]
Sure, I will add that to the pr later today.

**Comment by ajnyga** [comment]
@NateWr added that suggestion: #3542

**Comment by ajnyga** [comment]
@asmecher this would be ready for review as well. https://github.com/pkp/pkp-lib/pull/3542

**Comment by NateWr** [comment]
FYI, I've taken over this PR and I'll roll it into changes related to #3575.

**Comment by NateWr** [comment]
This issue was address in a PR merged with #3575.
